lib/infrastructure_collector.rb

```diff
@@ -74,38 +76,107 @@ def run
+        cluster = @clusters[host[:cluster]] || (@clusters[cluster_pid] if cluster_pid)
+        host[:cluster_platform_id] =  (host[:cluster] =~ /domain/).present? ? host[:cluster] : nil
+        host[:cluster] = cluster ? cluster[:name] : nil
+        host[:host_bus_adapters] = convert_speed_mbits_to_speed_bits_per_second(@host_bus_adapters)[host[:host_bus_adapters]]
+        host[:nics] = convert_speed_mbits_to_speed_bits_per_second(@nics)[host[:nics]]
+        @host_objects[key] = Host.new(host)
+      end
+      data_centers_hash.each do |platform_id, properties|
+        host_ids = hosts_for_datacenter(properties[:hostFolder])
+
+        properties[:hosts] = @host_objects.select { |k| host_ids.include?(k) }.values
+        properties[:networks] = properties[:network].map { |network| Network.new(name: network) } unless properties[:network].blank?
+        properties[:volumes] = properties[:datastores].map { |ds_moref| Volume.new(@infrastructure_volumes[ds_moref]) } unless properties[:datastores].blank?
+        properties.delete(:hostFolder)
+        properties.delete(:network)
+        properties.delete(:datastores) # Datastores from vSphere are referred to as 'volumes' in our domain
+        properties[:tags] = 'vmware-collector'
+        @local_inventory[platform_id] = updated_infrastructure(properties)
+      end
+      @local_inventory.save
     end
-
-    @local_inventory.save
-
     logger.debug 'Generating vSphere session activity with currentTime request'
     VSphere.wrapped_vsphere_request { VSphere.session.serviceInstance.CurrentTime }
   end
 
   private
 
+  def updated_infrastructure(properties)
+    infrastructure = Infrastructure.where(platform_id: properties[:platform_id])
+    if infrastructure.present?
+      current_infrastructure = infrastructure.first
+      update_fields_for(current_infrastructure,properties)
+    else
+      Infrastructure.new(properties)
+    end
+  end
+
+  def update_fields_for(current_infrastructure,properties)
+    BASIC_INFRASTRUCTURE_FIELDS.each{|field| current_infrastructure[field] = properties[field] }
+    current_infrastructure[:networks] = properties[:networks] ? build_updated_networks(properties) : []
+    current_infrastructure[:volumes] = properties[:volumes] ? build_updated_volumes(properties) : []
+    current_infrastructure.record_status = 'updated'
+    current_infrastructure = update_hosts(properties,current_infrastructure)
+    current_infrastructure.save
+    current_infrastructure
+  end
+
+  def update_hosts(properties,current_infrastructure)
+    properties[:hosts].each do |h|
+      host = current_infrastructure.hosts.where(platform_id: h[:platform_id]).first
+      host.present? ? update_host(host,h) : current_infrastructure.hosts << h
+    end
+    current_infrastructure.hosts = current_infrastructure.hosts.map{ |h| h if (properties[:hosts].map &:platform_id).include?h[:platform_id]} if !properties[:hosts].empty?
```

Comment:
What do you think if we switch this back to a regular if/else syntax instead of a single long line? (also a multi-line block for the map would be nice as well)